src/ir/diagnostic.cc

```diff
@@ -0,0 +1,246 @@
+  (*this)->diagnostics.push_back(diagnostic);
+}
+
+/*! \brief Emit a diagnostic. */
+void DiagnosticContext::EmitFatal(const Diagnostic& diagnostic) {
+  Emit(diagnostic);
+  Render();
+}
+
+/* Default Terminal Renderer. */
+static const char* DEFAULT_RENDERER = "diagnostics.DefaultRenderer";
+
+DiagnosticContext DiagnosticContext::Default(const IRModule& module) {
+  auto pf = tvm::runtime::Registry::Get(DEFAULT_RENDERER);
+  ICHECK(pf != nullptr)
+      << "Can not find registered function for " << DEFAULT_RENDERER << "." << std::endl
+      << "Either this is an internal error or the default function was overloaded incorrectly.";
+  auto default_render = tvm::runtime::TypedPackedFunc<DiagnosticRenderer()>(*pf);
+  return DiagnosticContext(module, default_render());
+}
+
+std::ostream& EmitDiagnosticHeader(std::ostream& out, const Span& span, DiagnosticLevel level,
+                                   std::string msg) {
+  rang::fg diagnostic_color;
+  std::string diagnostic_type;
+
+  switch (level) {
+    case DiagnosticLevel::kWarning: {
+      diagnostic_color = rang::fg::yellow;
+      diagnostic_type = "warning";
+      break;
+    }
+    case DiagnosticLevel::kError: {
+      diagnostic_color = rang::fg::red;
+      diagnostic_type = "error";
+      break;
+    }
+    case DiagnosticLevel::kBug: {
+      diagnostic_color = rang::fg::blue;
+      diagnostic_type = "bug";
+      break;
+    }
+    case DiagnosticLevel::kNote: {
+      diagnostic_color = rang::fg::reset;
+      diagnostic_type = "note";
+      break;
+    }
+    case DiagnosticLevel::kHelp: {
+      diagnostic_color = rang::fg::reset;
+      diagnostic_type = "help";
+      break;
+    }
+  }
+
+  out << rang::style::bold << diagnostic_color << diagnostic_type << ": " << rang::fg::reset << msg
+      << std::endl
+      << rang::fg::blue << " --> " << rang::fg::reset << rang::style::reset
+      << span->source_name->name << ":" << span->line << ":" << span->column << std::endl;
+
+  return out;
+}
+
+/*! \brief Generate an error message at a specific line and column with the
+ * annotated message.
+ *
+ * The error is written directly to the `out` std::ostream.
+ *
+ * \param out The output ostream.
+ * \param line The line at which to report a diagnostic.
+ * \param line The column at which to report a diagnostic.
+ * \param msg The message to attach.
+ */
+void ReportAt(const DiagnosticContext& context, std::ostream& out, const Span& span,
+              const Diagnostic& diagnostic) {
+  auto it = context->module->source_map->source_map.find(span->source_name);
+
+  // If the source name is not in the current source map, sources were not annotated.
+  if (it == context->module->source_map->source_map.end()) {
+    LOG(FATAL) << "The source maps are not populated for this module."
+               << "Please use `tvm.relay.transform.AnnotateSpans` to attach source maps for error "
+                  "reporting."
```

Comment:
```suggestion
    LOG(FATAL) << "The source maps are not populated for this module. "
               << "Please use `tvm.relay.transform.AnnotateSpans` to attach source maps for error "
                  "reporting. "
```

Reply:
I don't think need an extra space here, this implicitly writes a newline right after the message.